File: park-control/scripts/replay.js
/*
 * Replay a synthetic car lifecycle against the running listener, so you can
 * exercise the state machine before the simulator is wired up.
 *
 *   node scripts/replay.js
 *
 * Events use the exact shapes from the webhook documentation.
 */
const BASE = process.env.TARGET ?? 'http://127.0.0.1:4000/webhook';
const PLATE = process.env.PLATE ?? 'WCT 759';

let seq = Number(process.env.START_SEQ ?? 9000);
const uuid = () => crypto.randomUUID();
const stamp = () => new Date().toISOString().slice(0, 19).replace('T', ' ');

async function send(event) {
  const body = { ...event, EventId: uuid(), SequenceId: seq++, ServerDateTime: stamp() };
  const res = await fetch(BASE, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(body),
  });
  const text = await res.text();
  console.log(`${String(body.SequenceId).padEnd(6)} ${body.EventClass.padEnd(22)} -> ${res.status} ${text}`);
  return res;
}

const carEvent = (spotName, spotType, direction, extra = {}) => ({
  EventClass: 'car_spot_action',
  CarPlateNumber: PLATE,
  SpotName: spotName,
  SpotType: spotType,
  CarType: 'Normal',
  Direction: direction,
  PlannedParkingDurationInMinutes: '0',
  ...extra,
});

const wait = (ms) => new Promise((r) => setTimeout(r, ms));

async function main() {
  console.log(`replaying a full car lifecycle for ${PLATE} against ${BASE}\n`);

  await send({ EventClass: 'test_webhook' });

  await send(carEvent('ENTRY1', 'EntrySpot', 'CarIn'));
  await wait(300);
  await send(carEvent('ENTRY1', 'EntrySpot', 'CarOut'));
  await wait(300);

  await send(carEvent('S3', 'Park', 'CarIn', { PlannedParkingDurationInMinutes: '2' }));
  console.log('\n... car is parked, waiting 3s to accrue billable time ...\n');
  await wait(3000);

  await send(carEvent('S3', 'Park', 'CarOut', { PlannedParkingDurationInMinutes: '2' }));
  await wait(300);
  await send(carEvent('EXIT_EXIT', 'ExitSpot', 'CarIn'));
  await wait(300);

  // Honest payment. Flip to a wrong number to watch fraud detection fire.
  await send({
    EventClass: 'payment_made',
    CarPlateNumber: PLATE,
    Amount: process.env.AMOUNT ?? '1.00',
    Reason: 'Car Payment',
  });
  await wait(300);

  await send(carEvent('EXIT_EXIT', 'ExitSpot', 'CarOut'));

  console.log('\n--- extras ---');
  await send({
    EventClass: 'component_broken',
    Type: 'BarrierGate',
    Name: 'gateA',
    FineAmount: '10.00',
  });
  await send({
    EventClass: 'penalty',
    Reason: 'BarrierGate: Cannot Operate if it is Broken or under Maintenance.',
    FineAmount: '10',
    Type: 'BarrierGate',
    ComponentName: 'gateA',
  });
  await send({
    EventClass: 'carbon_monoxide_event',
    ZoneName: 'ZONE1',
    CarbonMonoxideLevel: 63.564693,
    DangerLevel: 'Mid',
  });

  // Duplicate delivery: must be rejected as already seen.
  console.log('\n--- duplicate EventId check ---');
  const dupe = {
    EventClass: 'gate_action',
    Name: 'gateA',
    Action: 'Open',
    EventId: 'fixed-id-for-dupe-test',
    SequenceId: seq++,
    ServerDateTime: stamp(),
  };
  for (let i = 0; i < 2; i++) {
    const res = await fetch(BASE, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(dupe),
    });
    console.log(`  attempt ${i + 1} -> ${await res.text()}`);
  }

  // Sequence gap: should be recorded, not silently swallowed.
  console.log('\n--- sequence gap check ---');
  seq += 25;
  await send({ EventClass: 'gate_action', Name: 'gateA', Action: 'Closed' });

  console.log('\nnow check:  curl http://127.0.0.1:4000/health');
}

main().catch((err) => {
  console.error('replay failed:', err.message);
  console.error('is the listener running?  npm start');
  process.exit(1);
});

File: scripts/replay.py
"""Drive a synthetic car lifecycle at the dispatcher, with no simulator running.

    python -m scripts.replay

Exercises the full path plus the guards that protect the score: duplicate
delivery, sequence gaps, penalties, CO events and a fake payment.

    AMOUNT=0.01 python -m scripts.replay     # watch fraud detection trip
"""
from __future__ import annotations

import json
import os
import time
import urllib.error
import urllib.request
import uuid
from datetime import datetime, timezone
from typing import Any

TARGET = os.environ.get("TARGET", "http://127.0.0.1:8080/webhooks/simulator")
PLATE = os.environ.get("PLATE", "WCT 759")
AMOUNT = os.environ.get("AMOUNT")

_seq = int(os.environ.get("START_SEQ", "9000"))


def _stamp() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")


def send(event: dict[str, Any], *, event_id: str | None = None) -> str:
    global _seq
    body = {
        **event,
        "EventId": event_id or str(uuid.uuid4()),
        "SequenceId": _seq,
        "ServerDateTime": _stamp(),
    }
    _seq += 1

    request = urllib.request.Request(
        TARGET,
        data=json.dumps(body).encode(),
        headers={"content-type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(request, timeout=10) as response:
            text = response.read().decode()
            status = response.status
    except urllib.error.HTTPError as exc:
        text, status = exc.read().decode(), exc.code
    except urllib.error.URLError as exc:
        raise SystemExit(f"cannot reach {TARGET}: {exc.reason}\nis the dispatcher running?")

    print(f"{body['SequenceId']:<6} {body['EventClass']:<22} -> {status} {text}")
    return text


def car(spot: str, spot_type: str, direction: str, **extra: Any) -> dict[str, Any]:
    return {
        "EventClass": "car_spot_action",
        "CarPlateNumber": PLATE,
        "SpotName": spot,
        "SpotType": spot_type,
        "CarType": os.environ.get("CAR_TYPE", "Normal"),
        "Direction": direction,
        "PlannedParkingDurationInMinutes": "0",
        **extra,
    }


def main() -> None:
    print(f"replaying a full lifecycle for {PLATE} against {TARGET}\n")

    send({"EventClass": "test_webhook"})

    send(car("ENTRY1", "EntrySpot", "CarIn"))
    time.sleep(0.3)
    send(car("ENTRY1", "EntrySpot", "CarOut"))
    time.sleep(0.3)

    send(car("S3", "Park", "CarIn", PlannedParkingDurationInMinutes="2"))
    print("\n... parked, waiting 3s to accrue billable time ...\n")
    time.sleep(3)

    send(car("S3", "Park", "CarOut", PlannedParkingDurationInMinutes="2"))
    time.sleep(0.3)
    send(car("EXIT_EXIT", "ExitSpot", "CarIn"))
    time.sleep(0.3)

    # Default pays exactly what a ~0.05 minute stay costs at 1/minute.
    amount = AMOUNT if AMOUNT is not None else "0.05"
    send({
        "EventClass": "payment_made",
        "CarPlateNumber": PLATE,
        "Amount": amount,
        "Reason": "Car Payment",
    })
    time.sleep(0.3)
    send(car("EXIT_EXIT", "ExitSpot", "CarOut"))

    print("\n--- component + penalty events ---")
    send({"EventClass": "component_broken", "Type": "BarrierGate",
          "Name": "gateA", "FineAmount": "10.00"})
    send({"EventClass": "penalty",
          "Reason": "BarrierGate: Cannot Operate if it is Broken or under Maintenance.",
          "FineAmount": "10", "Type": "BarrierGate", "ComponentName": "gateA"})
    send({"EventClass": "carbon_monoxide_event", "ZoneName": "ZONE1",
          "CarbonMonoxideLevel": 63.564693, "DangerLevel": "Mid"})

    print("\n--- duplicate EventId (second must be rejected) ---")
    fixed = "fixed-id-for-dupe-test"
    send({"EventClass": "gate_action", "Name": "gateA", "Action": "Open"}, event_id=fixed)
    send({"EventClass": "gate_action", "Name": "gateA", "Action": "Open"}, event_id=fixed)

    print("\n--- sequence gap (must be recorded) ---")
    global _seq
    _seq += 25
    send({"EventClass": "gate_action", "Name": "gateA", "Action": "Closed"})

    print("\nnow check:  curl http://127.0.0.1:8080/healthz")


if __name__ == "__main__":
    main()
